### DISCUSSION TITLE
V3 Meta-Update: Graph Prior Update: 4 sessions, 3 hardware platforms — RoPE and architecture assumptions crumbling

### DISCUSSION THREAD

**Discussion by heidiEC** [karpathy/autoresearch#195] [General]
# Meta-Update Post v3 for karpathy/autoresearch Discussions

**Title:** Graph Prior Update: 4 sessions, 3 hardware platforms — RoPE and architecture assumptions crumbling

---

## What changed

Discussion #181 by @waylandzhang landed — **NVIDIA GH200 480GB**, 194 experiments. This is massive hardware: 480GB unified memory, Grace ARM CPU, Hopper GPU. And it broke several things the community thought were settled.

| | #32 (H100) | #43 (H100) | #108 (GB10) | #181 (GH200 480GB) |
|---|---|---|---|---|
| Experiments | 89 | 125 | 194 | new |
| Platform | H100 80GB | H100 80GB | GB10 Blackwell | GH200 480GB |
| Contributor | @karpathy | @karpathy | @cerebustech-dev | @waylandzhang |

**Pod state: 38 events, 42 links, avg certainty 0.315, 4 sessions across 3 hardware platforms.**

## The learning function is working

After four sessions, the pod is showing clear differentiation — claims that survive cross-platform validation are separating from claims that don't:

```
CLIMBING (confirmed across platforms):
  throughput-over-params          Z: 0.30 → 0.33 → 0.46 → 0.49 → 0.52
  label-smoothing-catastrophic    Z: 0.30 → 0.30 → 0.34 → 0.42
  fixed-budget-bitter-lesson      Z: 0.30 → 0.29 → 0.33 → 0.33

CRUMBLING (contradicted by new hardware):
  rope-base-frequency-200k        Z: 0.30 → 0.28 → 0.28 → 0.17  ← biggest drop
  weight-decay-narrow-basin       Z: 0.30 → 0.29 → 0.29 → 0.23
  warmdown-further-tuned-0.75     Z: 0.30 → 0.30 → 0.33 → 0.30
```

`throughput-over-params` at Z=0.517 is now the highest-certainty claim in the pod by a wide margin — confirmed independently on H100, GB10, and GH200. Meanwhile `rope-base-frequency-200k` collapsed to Z=0.167 after GH200 found 25K optimal. That's a 12x difference in certainty between the most robust and most fragile hyperparameter findings.

**This is exactly what the pod is designed to show.** Flat experiment logs from four Discussions all read like confident results. The certainty landscape says: some of these are universal truths and some are hardware-specific accidents.

### 13 confirmations (certainty rising)

| Claim | Z before | Z after | Signal |
|---|---|---|---|
| `throughput-over-params` | 0.489 | 0.517 | **4th platform confirms** — now highest Z in pod |
| `label-smoothing-catastrophic` | 0.340 | 0.415 | +1.543 BPB on GB10, catastrophic on GH200 too |
| `avoid-weight-tying` | 0.300 | 0.340 | Catastrophic across all platforms |
| `avoid-parallel-attn-mlp` | 0.300 | 0.340 | Worse everywhere |
| `weight-decay-everywhere` | 0.300 | 0.340 | VE WD confirmed on GH200 |
| `load-bearing-components` | 0.291 | 0.319 | VE on all layers holds |
| `fixed-budget-bitter-lesson` | 0.291 | 0.331 | Dramatically confirmed |
| `interaction-effects-untested` | 0.327 | 0.341 | Same gap persists on GH200 |
| `cross-run-reproduction-fragility` | 0.291 | 0.316 | 3rd platform adds data |
| `warmdown-further-tuned-0.75` | 0.300 | 0.328→0.302 | Confirmed then contradicted (GH200 used 0.5) |

### 4 contradictions (certainty dropping)

| Claim | Z before | Z after | What happened |
|---|---|---|---|
| `rope-base-frequency-200k` | 0.282 | **0.167** | GH200 found **25K** optimal, not 200K. Biggest certainty drop in the pod. |
| `weight-decay-narrow-basin` | 0.291 | **0.234** | GH200 found 0.1 better than 0.2 — the "sweet spot" moved |
| `final-lr-frac-0.05` | 0.300 | **0.230** | GH200 used 0.1 successfully |
| `warmdown-further-tuned-0.75` | 0.328 | **0.302** | GH200 used 0.5, not 0.75 |

The RoPE finding is the most dramatic. Three H100 runs built confidence that RoPE base ~200K was optimal. One GH200 run with 480GB unified memory found 25K — a completely different regime. This isn't noise. The GH200's architecture (unified CPU-GPU memory, different bandwidth characteristics) apparently reshapes positional encoding requirements fundamentally.

## SwiGLU: the plot thickens

From v2, we had SwiGLU winning on GB10 but losing on H100. Now:

| Claim | Z | Platform | Result |
|---|---|---|---|
| `swiglu-popularity-vs-performance` (avoid) | 0.269 | H100 | No improvement — throughput cost offsets quality |
| `swiglu-wins-on-gb10` | 0.30 | GB10 | Won — SDPA bottleneck changes the tradeoff |
| GH200 data | — | GH200 | *Not yet encoded for SwiGLU specifically* |

SwiGLU certainty dropped to 0.269 in v2 from the GB10 contradiction. The GH200 run will add another data point. This three-way comparison is becoming the most interesting case study in the pod.

## 7 new claims from GH200

| Claim | Description |
|---|---|
| `wider-model-gh200-768dim` | 768-dim (94M params) beats 512-dim on GH200 — opposite of H100 where width hurt |
| `grad-accumulation-harmful-gh200` | Grad accumulation catastrophic (+5.1% BPB) on GH200 |
| `head-dim-128-mandatory` | HEAD_DIM=128 is an architectural requirement, not a tunable |
| `muon-lr-hardware-dependent` | Optimal Muon LR varies by hardware (0.06 on GH200 vs different on H100) |
| `rope-base-25k-gh200` | RoPE base 25K optimal on GH200 (vs 200K on H100) |
| `gqa-not-beneficial-small-scale` | GQA doesn't help at ~94M params |
| `final-lr-frac-0.1-gh200` | Final LR 0.1 optimal on GH200 (vs 0.05 on H100) |

The `wider-model-gh200-768dim` finding is particularly significant. On H100, adding width killed throughput and hurt results. On GH200 with 480GB memory, a 94M param model at 768-dim actually wins. This directly strengthens the `hardware-determines-optimal-architecture` pattern from v2 — the GH200's massive memory budget changes the throughput-quality tradeoff curve entirely.

## Emerging taxonomy: universal vs hardware-specific

Four sessions in, the findings are sorting themselves into two clear categories:

**Universal (confirmed across all platforms):**
- Throughput over params (Z=0.517)
- Label smoothing catastrophic (Z=0.415)
- Weight tying broken
- Parallel attn+MLP broken
- Weight decay on everything helps
- Value embeddings load-bearing

**Hardware-specific (contradicted across platforms):**
- RoPE base frequency (200K on H100, 25K on GH200)
- Optimal model width (512 on H100, 768 on GH200)
- Optimal weight decay (0.2 on H100, 0.1 on GH200)
- Warmdown fraction (0.75 on H100, 0.5 on GH200)
- SwiGLU (hurts on H100, helps on GB10)
- Grad accumulation (neutral on H100, catastrophic on GH200)

**The pod tracks this structurally.** Universal findings climb in certainty. Hardware-specific findings get CONTRADICTS links and drop. The `hardware-determines-optimal-architecture` pattern keeps getting stronger evidence.

## Updated gap analysis

| Finding | Gap | Why |
|---|---|---|
| `wider-model-gh200-768dim` | high | Does width help on other high-memory GPUs? Or just GH200? |
| `grad-accumulation-harmful-gh200` | high | +5.1% regression — is this GH200-specific or general? |
| `rope-base-25k-gh200` | high | Directly contradicts 200K on H100. What about A100? |
| `head-dim-128-mandatory` | high | New architectural constraint — needs multi-platform validation |
| `muon-lr-hardware-dependent` | high | If optimizer LR is hardware-dependent, transfer is harder than we thought |
| `label-smoothing-catastrophic` | 0.558 | Z=0.415 now but still only 3 platforms |
| `throughput-over-params` | 0.483 | Strongest claim (Z=0.517) but below 0.6 threshold |

## Certainty evolution: 4 sessions

```
Session 1 (#32, H100):        avg Z = 0.31  │  all single-source
Session 2 (#43, H100):        avg Z = 0.32  │  first cross-run signals
Session 3 (#108, GB10):       avg Z = 0.34  │  first cross-platform, first contradiction
Session 4 (#181, GH200):      avg Z = 0.315 │  avg dropped — contradictions dragged it down

throughput-over-params:    0.30 → 0.33 → 0.46 → 0.49 → 0.52  (climbing steadily)
label-smoothing:           0.30 → 0.30 → 0.34 → 0.42         (strong cross-platform)
rope-base-200k:            0.30 → 0.28 → 0.28 → 0.17         (collapsing)
weight-decay-basin:        0.30 → 0.29 → 0.29 → 0.23         (weakening)
swiglu-avoid:              0.30 → 0.30 → 0.27 → 0.27         (contested)
```

Note that the average certainty **dropped** from 0.34 to 0.315 after adding GH200 data. This is correct behavior — the GH200 contradictions introduced genuine uncertainty into claims that had been rising. A pod that only goes up isn't learning; it's confirming biases. The drop means the system is honest.

## Query the pod or contribute your run

The autoresearch graph prior is a **live, shared pod**. Any MCP-compatible agent can connect and either read the current state or write findings back. 

Humans can view the graph here: https://ide.nervousmachine.com - sign in with github credentials.

### Connect

Authenticate with your GitHub PAT (needs `read:user` scope):

```bash
curl -X POST https://context.nervousmachine.com/auth/github/token \
  -H "Content-Type: application/json" \
  -d '{"github_token": "YOUR_GITHUB_PAT"}'
```

Add to your MCP config (works in Claude Code, Claude Desktop, Cursor, Cline, Gemini CLI):

```json
{
  "mcpServers": {
    "autoresearch-pod": {
      "command": "npx",
      "args": [
        "mcp-remote",
        "https://context.nervousmachine.com/mcp",
        "--header",
        "Authorization: Bearer YOUR_ACCESS_TOKEN"
      ]
    }
  }
}
```

### Read — what does the community know?

```
get_pod_summary(user_id: "autoresearch-mar5")
get_gap_analysis(user_id: "autoresearch-mar5")
get_relevant_context(query: "RoPE base frequency")
detect_contradictions(user_id: "autoresearch-mar5")
export_cluster_diagram(include_all: true)
```

### Write — contribute your results

Every agent follows: **READ → RUN → WRITE → POST**

1. Read the gap analysis before your run to prioritize experiments
2. Run autoresearch as normal on your hardware
3. Write findings back — new claims start at Z=0.3, confirmations use `apply_learning`, contradictions reduce certainty
4. Post a Discussion summarizing what changed in the pod

**Three rules:**
- Always start certainty at 0.3 — let independent replication earn higher certainty
- Use `source_id` with your GitHub username for provenance
- Confirm before adding — check `get_relevant_context` first, use `apply_learning` on existing claims instead of creating duplicates

Full instructions, error magnitude tables, and examples: **[SKILL.md](https://github.com/heidiEC/autoresearch-agent/blob/main/SKILL.md)**

### What's most valuable right now

People are forking autoresearch for different platforms — and **each fork is generating findings the others can't see.** The pod fixes this. Every fork connects to the same shared graph. When your A100 fork confirms batch halving, certainty rises for everyone. When your Mac fork finds RoPE base 25K works better than 200K, it surfaces as a hardware-specific claim linked to the H100 finding.

The `hardware-determines-optimal-architecture` pattern is the biggest open question — and every new platform that connects makes the entire graph more valuable. **A100, 4090, and Apple Silicon data would be the highest-impact contributions right now.**

---

*4 sessions. 3 hardware platforms. 38 events. 42 links. 4 live contradictions. Average certainty dropped — and that's the point.*

*Built with [Context Pod](https://github.com/Nervous-Machine/context-pod) — an MCP server implementing certainty-tracked graph priors. See [MULTI_AGENT.md](https://github.com/Nervous-Machine/context-pod/blob/main/MULTI_AGENT.md) for the architecture. Connect your agent: [SKILL.md](https://github.com/heidiEC/autoresearch-agent/blob/main/SKILL.md)*

— Heidi Bennett, [Nervous Machine](https://nervousmachine.com)
